### ISSUE TITLE
Implement dynamical decoupling transformer

### ISSUE THREAD

**Issue by babacry** [quantumlib/Cirq#6582]
**Enable dynamical decoupling operations insertion**

**Acceptance criteria - Users are able to use the transformer to insert dynamical decoupling sequences based on their preferences**

**Comment by babacry** [comment]
Suggestions from snichet@

1. [Feature 1](https://github.com/quantumlib/Cirq/pull/6515/commits/3a6a9b4decdf6e9675e66175c6cb5e61975f76ae#r1538340780) Might be nice to have an option only to insert DD gates on moments that only have single qubits gates. On the hardware, the calibrations for the gates do not include calibrating single qubits and two qubits gates concurrently.
2. [Feature 2](https://github.com/quantumlib/Cirq/pull/6515/commits/3a6a9b4decdf6e9675e66175c6cb5e61975f76ae#r1538359878) It seems like this is only inserting X gates on consecutive idle moments. I think the next step would be to divide the circuit into moments that are Clifford, then insert X gates on all idle single qubit moments (and perhaps also idle 2 qubit moments, but the caveat is my comment above). Then use the fact that Clifford circuits can be quickly and efficiently simulated to put the correct Clifford gate at the end of the group of moments, to inverse all the previous inserted gates. For CZ gates, this should just be a X, Y or Z gate, i believe.

The features will be considered after #6515

**Comment by babacry** [comment]
#6675 implemented features above.

**Comment by babacry** [comment]
#6718

1. Insert for each Clifford piece.
2. Insert for single-qubit consecutive moments.
